src/meta/server_state.cpp

```diff
@@ -3282,7 +3283,93 @@ void server_state::set_max_replica_count(configuration_set_max_replica_count_rpc
              response.old_max_replica_count,
              new_max_replica_count);
 
-    do_update_max_replica_count(app, rpc);
+    set_max_replica_count_env_updating(
+        app, rpc, [this, app, rpc]() mutable { do_update_max_replica_count(app, rpc); });
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void server_state::set_max_replica_count_env_updating(
+    std::shared_ptr<app_state> &app,
+    configuration_set_max_replica_count_rpc rpc,
+    max_replica_count_env_callback on_max_replica_count_env_updating)
+{
+    zauto_write_lock l(_lock);
+
+    auto iter = app->envs.find(replica_envs::UPDATE_MAX_REPLICA_COUNT);
+    if (iter != app->envs.end()) {
+        std::vector<std::string> args;
+        utils::split_args(iter->second.c_str(), args, ';');
+        if (args[0] == "updating") {
+            auto &response = rpc.response();
+            response.err = ERR_OPERATION_DISABLED;
+            response.hint_message = fmt::format("max_replica_count of app({}) is being updated, "
+                                                "thus this request would be rejected",
+                                                app->app_name);
+            derror_f("failed to set max_replica_count: app_name={}, app_id={}, error_code={}, "
+                     "hint_message={}",
+                     app->app_name,
+                     app->app_id,
+                     response.err.to_string(),
+                     response.hint_message);
+            return;
+        }
+    }
+
+    const auto new_max_replica_count = rpc.request().max_replica_count;
+    const auto old_max_replica_count = rpc.response().old_max_replica_count;
+
+    ddebug_f("ready to update local env of max_replica_count: app_name={}, app_id={}, "
+             "old_max_replica_count={}, new_max_replica_count={}, {}={}",
+             app->app_name,
+             app->app_id,
+             old_max_replica_count,
+             new_max_replica_count,
+             replica_envs::UPDATE_MAX_REPLICA_COUNT,
+             app->envs[replica_envs::UPDATE_MAX_REPLICA_COUNT]);
+
+    // Update local env first, since another thread that is trying to update max_replica_count
+    // will be blocked by write lock.
+    app->envs[replica_envs::UPDATE_MAX_REPLICA_COUNT] =
+        fmt::format("updating;{}", new_max_replica_count);
```

Comment:
If remote update is executed first, it will lead to inconsistency since the process of local and remote update is not atomic. 

Suppose there are 2 requests R1 and R2. R1 tries to set `max_replica_count` to 2, R2 tries to set `max_replica_count` to 3. If remote update is executed first,  we can assume that R1 finishes remote update fist, and then R2 finishes remote update. Thus remote env will be `updating;3`. However, the callback that sets local env for R1 is run first, thus local env will be set to `updating;2`. Local and remote env will become inconsistent.